### PR TITLE
Support Scaled matmul and scaled dot_general in jax.nn

### PR DESCRIPTION
**Pull request by wenscarl**
This PR introduces two new nn APIs:

1. `scaled_matmul` – Performs matrix multiplication with scaling factors.
2. `scaled_dot_general` – Implements a generalized dot product operation with quantization, matmul and dequantization.
Currently, only the e8m0fnu type scaling factor is supported.
@kaixih @nouiz

### CODE REVIEW

**Comment by kaixih** [comment]
@IvyZX @sdasgup3 can you help review or find other reviewers?

**Comment by IvyZX** [comment]
PTAL test errors - looks like `e8m0fnu` is a type that neither Numpy or JAX supports. I'm not sure if we should adopt this type in JAX.

Also, is it our common practice to add purely-cudnn APIs to `jax.nn`? Looks like in https://github.com/jax-ml/jax/pull/21371 an XLA implementation is provided as an alternative for users without cudnn.

Adding maintainers who know more about JAX's lower level common practices: @hawkinsp @superbobry

**Comment by kaixih** [comment]
> PTAL test errors - looks like `e8m0fnu` is a not type that Numpy supports. I'm not sure if we should adopt this type in JAX.

This dtype should already in the JAX (see [here](https://github.com/jax-ml/jax/blob/531d80dc72a2c78030aa47a11ffbc07ce4f03466/jax/_src/dtypes.py#L97)). The dtype is first defined in ml_dtype since [0.5.0](https://github.com/jax-ml/ml_dtypes/releases/tag/v0.5.0) and imported to JAX. 

> Also, is it our common practice to add purely-cudnn APIs to `jax.nn`? Looks like in #21371 an XLA implementation is provided as an alternative for users without cudnn.

The `scaled_matmul` and `scaled_dot_general` are supposed to be general and not cudnn-specific. Yes, both of them call into the `cudnn_scaled_matmul`, which appears to be purely-cudnn API. But `cudnn_scaled_matmul` will actually fall back to XLA impl if cudnn doesn't support the case (see [this XLA pass](https://github.com/openxla/xla/pull/21800/files)). So, that explains why we don't see the XLA impl in the python code.

**Comment by sdasgup3** [comment]
cc @reedwm

File: jax/_src/cudnn/scaled_matmul_stablehlo.py [hunk 1]
```diff
@@ -0,0 +1,730 @@
...
+  return (core.ShapedArray(output_shape, preferred_element_type),)
+
+
+_scaled_matmul_p = core.Primitive("scaled_matmul")
+_scaled_matmul_p.multiple_results = True
+_scaled_matmul_p.def_impl(partial(xla.apply_primitive, _scaled_matmul_p))
+_scaled_matmul_p.def_abstract_eval(_scaled_matmul_abstract)
+
+
+mlir.register_lowering(
```
Comment (IvyZX) [inline jax/_src/cudnn/scaled_matmul_stablehlo.py:117]: See [CI test failures](https://github.com/jax-ml/jax/actions/runs/13423839455/job/37506043305?pr=26345) - this code only runs on GPU now.

You will need to either: 
* add lowering fallback to XLA for other platforms like CPU and TPU, or
* explicitly say your implementation only works for certain platforms and prevent invalid usages: platform-check or require an explicit platform arg in the API (like this [cudnn implementation](https://github.com/jax-ml/jax/blob/main/jax/_src/nn/functions.py#L1019)), and disable tests on other platforms.

File: jax/_src/cudnn/scaled_matmul_stablehlo.py [hunk 2]
```diff
@@ -0,0 +1,730 @@
...
+  return clipped_x.astype(q_dtype)
+
+def quantize_dequantize(x, q_dtype, scale, compute_dtype):
+  qx = quantize_to_qtype(x, q_dtype, compute_dtype, scale)
+  out = qx.astype(x.dtype) * jnp.broadcast_to(
+      jnp.asarray(scale, dtype=x.dtype), qx.shape
+  )
+  return out
+
+def generate_quantized_tensors(
```
Comment (IvyZX) [inline jax/_src/cudnn/scaled_matmul_stablehlo.py:522]: This function is a test utility to generate random numbers. Can you move it to tests? You can replicate them in both tests. 

Same for `quantize_dequantize` above it.

File: tests/scaled_matmul_stablehlo_test.py [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+
+      j_ref = jax.jit(
+          partial(
+              jax.lax.dot_general,
+              dimension_numbers=(([2], [2]), ([0], [0])),
+          ),
+          in_shardings=input_shardings[:2],
+      )
+
+      out = j_scaled_matmul(*args)
```
Comment (IvyZX) [inline tests/scaled_matmul_stablehlo_test.py:222]: You may want to make sure the output is in the expected sharding using [`sharding.is_equivalent_to`](https://docs.jax.dev/en/latest/jax.sharding.html#jax.sharding.Sharding.is_equivalent_to).

Same for sharded tests below.

**Review by IvyZX — CHANGES_REQUESTED** [review]
Please also make sure all CI tests pass.

File: jax/_src/cudnn/scaled_matmul_stablehlo.py [hunk 4]
```diff
@@ -0,0 +1,730 @@
...
+
+# Given the inputs already sharded as
+#   ([B], M, K1), ([B], N, K2)
+# We define the following rule to apply necessary AllGather based on
+# "Input specs", and to define the "Output spec".
+# 1. If K1 == K2 != None and N == None:
+#   - Input spec : ([B], M, K1), ([B], None, K2)
+#   - Output spec: ([B], M, None) -> AllReduce -> ([B], M, None)
+# 2. If K1 == K2 != None and M == N != None:
+#   - Input spec : ([B], M, K1), ([B], None, K2)
```
[the remark this answers is not in the document]
Comment (wenscarl) [reply jax/_src/cudnn/scaled_matmul_stablehlo.py:136]: It's ([B], None, K2) for this case. Essentially we make the N dim None to perform a all-gather to the input before feeding into custom-call.

**Review by IvyZX — CHANGES_REQUESTED** [review]
Thank you for fixing all the tests! Only a few minor nits now.

**Comment by sdasgup3** [comment]
>> Implements a generalized dot product operation with quantization, matmul and dequantization

IIUC, the PR expresses the quantization and dequantization in StableHLO using their underlying math and scaled dot as custom-call `__op$block_scaled_dot`. Can you please confirm if there already exist a mapping to map this cc to related cudnn API.

**Comment by wenscarl** [comment]
> > > Implements a generalized dot product operation with quantization, matmul and dequantization
> 
> IIUC, the PR expresses the quantization and dequantization in StableHLO using their underlying math and scaled dot as custom-call `__op$block_scaled_dot`. Can you please confirm if there already exist a mapping to map this cc to related cudnn API.

Yes. https://github.com/openxla/xla/pull/21800 added such cc.

**Review by IvyZX — APPROVED** [review]

**Review by IvyZX — CHANGES_REQUESTED** [review]
Hi - to pass the internal test, you need to add the Apache License header to the two new python files you created. Just copy it from any existing file.
I can try merging again after that.

**Comment by wenscarl** [comment]
> Hi - to pass the internal test, you need to add the Apache License header to the two new python files you created. Just copy it from any existing file.
> I can try merging again after that.

Done.

**Review by IvyZX — APPROVED** [review]
